lib/cache_swift.c

```diff
@@ -0,0 +1,518 @@
+    if (swift_err != SCERR_SUCCESS) {
+        keystone_end(conn->keystone_context);
+        ctx->set_error(ctx, 500, "failed to swift_start()");
+        return;
+    }
+
+    mapcache_swift_authenticate(ctx, cache, conn);
+    if (GC_HAS_ERROR(ctx)) {
+        keystone_end(conn->keystone_context);
+        swift_end(conn->swift_context);
+        return;
+    }
+
+    *conn_ = conn;
+}
+
+void mapcache_swift_connection_destructor(void *conn_) {
+    struct swift_connection *conn = (struct swift_connection *)conn_;
+
+    keystone_end(conn->keystone_context);
+    swift_end(conn->swift_context);
+    free(conn->keystone_context);
+    free(conn->swift_context);
+    free(conn);
+}
+
+static mapcache_pooled_connection* _swift_get_connection(mapcache_context *ctx, mapcache_cache_swift *cache, mapcache_tile *tile)
+{
+  mapcache_pooled_connection *pc;
+  struct swift_conn_params params;
+
+  params.cache = cache;
+
+  pc = mapcache_connection_pool_get_connection(ctx,cache->cache.name,mapcache_swift_connection_constructor,
+          mapcache_swift_connection_destructor, &params);
+
+  return pc;
+}
+
+static int _mapcache_cache_swift_has_tile(mapcache_context *ctx, mapcache_cache *pcache, mapcache_tile *tile) {
+    mapcache_pooled_connection *pc;
+    mapcache_cache_swift *cache = (mapcache_cache_swift*) pcache;
+    struct swift_connection *conn;
+    enum swift_error err;
+    int exists;
+    int error = MAPCACHE_FALSE;
+    char *container;
+    char *key;
+
+    key = mapcache_util_get_tile_key(ctx, tile, cache->key_template, " \r\n\t\f\e\a\b", "#");
+    if (GC_HAS_ERROR(ctx)) {
+        return MAPCACHE_FALSE;
+    }
+
+    if(strchr(cache->container_template,'{')) {
+      container = mapcache_util_get_tile_key(ctx, tile, cache->container_template, " \r\n\t\f\e\a\b", "#");
+    } else {
+      container = cache->container_template;
+    }
+
+    pc = _swift_get_connection(ctx, cache, tile);
+
+    if (GC_HAS_ERROR(ctx)) {
+        return MAPCACHE_FALSE;
+    }
+    conn = pc->connection;
+
+    err = swift_set_container(conn->swift_context, container);
+    if (err != SCERR_SUCCESS) {
+        ctx->set_error(ctx, 500, "swift: failed to set container %s: %d", container, err);
+    }
+    err = swift_set_object(conn->swift_context, key);
+    if (err != SCERR_SUCCESS) {
+        ctx->set_error(ctx, 500, "swift: failed to set object %s: %d", key, err);
+    }
+
+    /* retry loop */
+    for (int i = 0; i < 1; ++i) {
+        err = swift_has(conn->swift_context, &exists);
+        if (err == SCERR_AUTH_FAILED) {
+            mapcache_swift_authenticate(ctx, cache, conn);
+            if (GC_HAS_ERROR(ctx)) {
+                break;
+            }
+            continue;
+        } 
+        else {
+            if (err == SCERR_SUCCESS && exists) {
+                error = MAPCACHE_TRUE;
+            }
+            break;
+        }
+    }
+
+    mapcache_connection_pool_release_connection(ctx, pc);
```

Comment:
This should be called before all exit paths (i.e. it is currently not called on error condition exits)

Reply:
Right, I'll need to check those as-well.